File: src/main/java/ca/uwindsor/analyzing/KeyTermsFilter.java
package ca.uwindsor.analyzing;

import ca.uwindsor.common.TermsCollection;
import org.apache.lucene.analysis.TokenFilter;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;

import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Deque;

/**
 * Filter for key terms.
 */
public class KeyTermsFilter extends TokenFilter
{
    /**
     * Required value for the tokenizing.
     */
    private final CharTermAttribute termAttr = addAttribute(CharTermAttribute.class);

    /**
     * Buffer for building term pairs.
     */
    private final Deque<String> buffer = new ArrayDeque<>();

    /**
     * The terms.
     */
    private final TermsCollection termsCollection;

    /**
     * Start with an existing collection.
     *
     * @param input      The token stream input.
     * @param collection The starting collection.
     */
    public KeyTermsFilter(TokenStream input, Collection<String> collection)
    {
        super(input);
        termsCollection = new TermsCollection(collection);
    }

    /**
     * See if we should increment the number of tokens.
     *
     * @return True if a key term was matched, false otherwise.
     * @throws IOException Reading the input fails.
     */
    @Override
    public boolean incrementToken() throws IOException
    {
        // Loop until we have a match or are at the end of the file.
        while (true)
        {
            // If we have nothing in the buffer, load the next term.
            if (buffer.isEmpty())
            {
                // If there are no more tokens to load we are done.
                if (!input.incrementToken())
                {
                    return false;
                }

                // Add the term to our buffer.
                buffer.addLast(termAttr.toString());
            }

            // Combine any buffered phrases.
            String currentPhrase = String.join(" ", buffer);

            // Check if this is a term we wish to match.
            if (termsCollection.Contains(currentPhrase, false))
            {
                termAttr.setEmpty().append(currentPhrase);
                buffer.clear();
                return true;
            }

            // Check if there is any way this buffer could be a term.
            if (termsCollection.SetStartsWith(currentPhrase, false) == null)
            {
                // This is not a match and no potential for it to be part of a larger one so clear it.
                buffer.removeFirst();
                continue;
            }

            // Load the next token if this could be a term.
            if (input.incrementToken())
            {
                buffer.addLast(termAttr.toString());
                continue;
            }

            // Otherwise, there are no more tokens so stop.
            buffer.clear();
            return false;
        }
    }

    /**
     * Reset the buffer.
     *
     * @throws IOException Any errors from resetting.
     */
    @Override
    public void reset() throws IOException
    {
        super.reset();
        buffer.clear();
    }
}

File: Lucene/Analyzing/KeyTermsFilter.cs
using Lucene.Net.Analysis;
using Lucene.Net.Analysis.TokenAttributes;
using SearchEngine.Lucene.Utility;

namespace SearchEngine.Lucene.Analyzing;

/// <summary>
/// Filter to look for only key terms.
/// </summary>
public sealed class KeyTermsFilter : TokenFilter
{
    /// <summary>
    /// Required to help with building terms.
    /// </summary>
    private readonly ICharTermAttribute _termAttr;
    
    /// <summary>
    /// Buffered terms to try and build a larger key terms.
    /// </summary>
    private readonly LinkedList<string> _buffer = new();
    
    /// <summary>
    /// The key terms.
    /// </summary>
    private readonly TermsCollection _keyTerms;
    
    /// <summary>
    /// Create the filter.
    /// </summary>
    /// <param name="input">The token stream.</param>
    /// <param name="collection">The key terms.</param>
    internal KeyTermsFilter(TokenStream input, ICollection<string> collection) : base(input)
    {
        _keyTerms = new(collection);
        _termAttr = AddAttribute<ICharTermAttribute>();
    }

    /// <summary>
    /// Look for the next token.
    /// </summary>
    /// <returns>True if a key term has been matched, false otherwise</returns>
    public override bool IncrementToken()
    {
        // Key terms can be multiple words long, so we may need to loop.
        while (true)
        {
            // If there is nothing in the buffer, we need to load more.
            if (_buffer.Count < 1)
            {
                // If there is nothing left to load, then we are done.
                if (!m_input.IncrementToken())
                {
                    return false;
                }
                
                // Add the next term to the buffer.
                _buffer.AddLast(_termAttr.ToString());
            }

            // See what our current term is.
            string currentPhrase = string.Join(" ", _buffer);

            // If this is a keyword, return true.
            if (_keyTerms.Contains(currentPhrase, false))
            {
                _termAttr.SetEmpty().Append(currentPhrase);
                _buffer.Clear();
                return true;
            }

            // Otherwise, if there is no potential that this is a keyword, remove the first portion.
            if (_keyTerms.CollectionStartsWith(currentPhrase, false) == null)
            {
                _buffer.RemoveFirst();
                continue;
            }

            // Try and add another portion to the key term.
            if (m_input.IncrementToken())
            {
                _buffer.AddLast(_termAttr.ToString());
                continue;
            }
            
            // Otherwise, there is nothing left, so clear the buffer.
            _buffer.Clear();
            return false;
        }
    }

    /// <summary>
    /// Reset to clear the buffer.
    /// </summary>
    public override void Reset()
    {
        base.Reset();
        _buffer.Clear();
    }
}